src/librustc_mir/transform/cs_propagate.rs

```diff
@@ -0,0 +1,425 @@
+            // Make sure case like `var1 = var1 {op} x` does not make our knowledge incorrect.
+            Rvalue::BinaryOp(..) | Rvalue::CheckedBinaryOp(..) | Rvalue::UnaryOp(..) => {
+                lat.top(lval);
+            }
+            // Cannot be handled
+            Rvalue::Box(_) => { lat.top(lval); }
+            // Not handled, but could be. Disaggregation helps to not bother with this.
+            Rvalue::Aggregate(..) => { lat.top(lval); }
+            // Not handled, invalidate any knowledge about any variables used by this. Dangerous
+            // stuff and other dragons be here.
+            Rvalue::InlineAsm { ref outputs, ref inputs, asm: _ } => {
+                lat.top(lval);
+                for output in outputs { lat.top(output); }
+                for input in inputs {
+                    if let Operand::Consume(ref lval) = *input { lat.top(lval); }
+                }
+            }
+        };
+        lat
+    }
+
+    fn term(&self, t: &Terminator<'tcx>, mut lat: Self::Lattice)
+    -> Self::TerminatorReturn
+    {
+        let span = t.source_info.span;
+        let succ_count = t.successors().len();
+        let bool_const = |b: bool| Either::Const(Constant {
+            span: span,
+            ty: self.tcx.mk_bool(),
+            literal: Literal::Value { value: ConstVal::Bool(b) },
+        });
+        match t.kind {
+            TerminatorKind::If { cond: Operand::Consume(ref lval), .. } => {
+                let mut falsy = lat.clone();
+                falsy.insert(lval, bool_const(false));
+                lat.insert(lval, bool_const(true));
+                vec![lat, falsy]
+            }
+            TerminatorKind::SwitchInt { ref discr, ref values, switch_ty, .. } => {
+                let mut vec: Vec<_> = values.iter().map(|val| {
+                    let mut branch = lat.clone();
+                    branch.insert(discr, Either::Const(Constant {
+                        span: span,
+                        ty: switch_ty,
+                        literal: Literal::Value { value: val.clone() }
+                    }));
+                    branch
+                }).collect();
+                vec.push(lat);
+                vec
+            }
+            TerminatorKind::Drop { ref location, .. } => {
+                lat.remove(location);
+                vec![lat; succ_count]
+            }
+            TerminatorKind::DropAndReplace { ref location, ref unwind, ref value, .. } => {
+                match *value {
+                    Operand::Consume(ref lval) => {
+                        lat.remove(location);
+                        lat.remove(lval);
+                    },
+                    Operand::Constant(ref cnst) => {
+                        lat.insert(location, Either::Const(cnst.clone()));
+                    }
+                }
+                if unwind.is_some() {
+                    let mut unwind = lat.clone();
+                    unwind.remove(location);
```

Comment:
`location` ends up being `value` in both cases - that's the point of `DropAndReplace`.